Abinit+wannier90 fix

downloadico opened the pull request:
This PR allows abinit to be built with the wannier90 library.

downloadico committed:
wannier90: add versions 3.0.0 and 3.1.0 and 'shared variant'

Added versions 3.0.0 and 3.1.0

Added shared variant

Added url_for_version function as versions less than 3 are from the
wannier.org site and versions 3 and up are from github.com

Added the MPI libraries to the list of libs substituted into the make.sys file
in place of @LIBS

Made it possible to build a shared object version of the library for versions
< 3 by filtering the src/Makefile.2 file (based off of the patch from a src rpm
from RHEL for version 2.0.1)

Create a modules directory in the install prefix root directory and copy the
Fortran .mod files there.

Set the MPIFC variable to the Spack Fortran MPI compiler wrapper.

downloadico committed:
abinit: added 'wannier90' variant  which enables building abinit with wannier90

Added wannier90 variant

Made abinit depend on the shared object ('shared') variant of
wannier90 if the wannier90 variant is selected

Add configure args for wannier90 libs, includes, and binaries and to
set MPIFC

set the dft-flavor to wannier90 when wannier90 is enabled and only
set the dft flavor to 'atompaw+libxc' if wannier90 is not selected

downloadico commented:
I'm getting strange failures related to, I think python versions (on the test machine that checks PRs)

becker33 commented on var/spack/repos/builtin/packages/abinit/package.py:
```diff
@@ -95,12 +100,25 @@ def configure_args(self):
+                 .format(spec['wannier90'].prefix.modules))
+            oapp('--with-wannier90-bins={0}'
+                 .format(spec['wannier90'].prefix.bin))
+            oapp('MPIFC={}/mpifc'.format(spec['mpi'].prefix.bin))
```
```suggestion
            oapp('MPIFC={0}/mpifc'.format(spec['mpi'].prefix.bin))
```

downloadico committed:
Update var/spack/repos/builtin/packages/abinit/package.py

Co-authored-by: Greg Becker <becker33@llnl.gov>

becker33 commented on var/spack/repos/builtin/packages/abinit/package.py:
```diff
@@ -95,12 +100,25 @@ def configure_args(self):
             oapp('--with-mpi-prefix={0}'.format(spec['mpi'].prefix))
             oapp('--enable-mpi=yes')
             oapp('--enable-mpi-io=yes')
+            if '~wannier90' in spec:
```
I don't think this should be restricted to when mpi is enabled. Shoudn't this be moved to an `else` block closing the `if +wannier90` in spec:` conditional above?

becker33 commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -61,10 +93,76 @@ def edit(self, spec, prefix):
+                                      '../wannier_prog.F90 $(LIBRARY)']),
+                            join_path(self.stage.source_path,
+                                      'src/Makefile.2'))
+                filter_file(' '.join(['../../postw90.x: \\$\\(OBJS_POST\\)',
+                                      '\\$\\(POSTDIR\\)postw90.F90']),
+                            ' '.join(['../../postw90.x: $(OBJS_POST)',
+                                     '$(POSTDIR)postw90.F90 $(LIBRARY)']),
+                            join_path(self.stage.source_path,
+                                      'src/Makefile.2'))
+                filter_file(
+                    ' '.join([
+                        '\\$\\(COMPILER\\) ../wannier_prog.F90',
+                        '\\$\\(LDOPTS\\) \\$\\(OBJS\\) \\$\\(LIBS\\)',
+                        '-o ../../wannier90.x']),
+                    ' '.join(['$(COMPILER) -I../obj ../wannier_prog.F90',
+                              '$(LDOPTS) -L../.. -lwannier',
+                              '-o ../../wannier90.x']),
+                    join_path(self.stage.source_path,
+                              'src/Makefile.2'))
+                filter_file(
+                    ' '.join(['\\$\\(COMPILER\\) \\$\\(POSTDIR\\)postw90.F90',
+                              '\\$\\(POSTOPTS\\) \\$\\(LDOPTS\\)',
+                              '\\$\\(OBJS_POST\\)',
+                              '\\$\\(LIBS\\) -o ../../postw90.x']),
+                    ' '.join(['$(COMPILER) -I../obj $(POSTDIR)postw90.F90',
+                              '$(POSTOPTS) $(LDOPTS) $(OBJS_POST)',
+                              '-L../.. -lwannier $(LIBS) -o ../../postw90.x']),
+                    join_path(self.stage.source_path,
+                              'src/Makefile.2'))
+
+                filter_file(
+                    ' '.join(['\\$\\(AR\\) \\$\\(ARFLAGS\\)',
+                              '\\$\\(LIBRARY\\)',
+                              '\\$\\(OBJS2\\) \\$\\(OBJS\\)']),
+                    ' '.join(['$(MPIF90) $(FCOPTS) -shared -o '
+                              '$(LIBRARY) $(OBJS2) $(OBJS) $(LIBS)']),
+                    join_path(self.stage.source_path,
+                              'src/Makefile.2'))
```
For all of these, you can use raw strings to avoid needing double backslashes.

e.g. `filter_file(r'\$...)`

adamjstewart replied:
Better yet, you can use `string=True` and you won't need to escape anything.

becker33 commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -76,10 +174,17 @@ def install(self, spec, prefix):
-            join_path(self.prefix.lib, 'libwannier.a')
-        )
+        if '+shared' in spec:
+            if '@3:' in spec:
```
```suggestion
            if spec.satisfies('@3:'):
```

becker33 commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -91,12 +196,18 @@ def install(self, spec, prefix):
-            join_path(self.stage.source_path, 'utility', 'w90pov', 'w90pov'),
-            join_path(self.prefix.bin, 'w90pov')
-        )
+        if '@:2.999.999' in spec:
```
```suggestion
        if spec.satisfies('@:2.99')
```

downloadico committed:
Update var/spack/repos/builtin/packages/wannier90/package.py

Co-authored-by: Greg Becker <becker33@llnl.gov>

downloadico committed:
Update var/spack/repos/builtin/packages/wannier90/package.py

Co-authored-by: Greg Becker <becker33@llnl.gov>

adamjstewart commented on var/spack/repos/builtin/packages/abinit/package.py:
```diff
@@ -95,12 +100,25 @@ def configure_args(self):
+                 .format(spec['wannier90'].prefix.modules))
+            oapp('--with-wannier90-bins={0}'
+                 .format(spec['wannier90'].prefix.bin))
+            oapp('MPIFC={0}/mpifc'.format(spec['mpi'].prefix.bin))
```
This needs to be in a `+mpi` check

adamjstewart commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -16,7 +17,11 @@ class Wannier90(MakefilePackage):
     homepage = 'http://wannier.org'
     url = 'http://wannier.org/code/wannier90-2.0.1.tar.gz'
+    # Newer versions are here
+    new_url = 'https://github.com/wannier-developers/wannier90/archive/v3.1.0.tar.gz'
```
This url doesn't appear to be used anywhere. I would replace the `url` above with this one so that `spack versions` and `spack checksum` will work correctly.

adamjstewart commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -26,9 +31,34 @@ class Wannier90(MakefilePackage):
+            targets = [
+                'wannier', 'post', 'lib', 'w90chk2chk', 'w90vdw', 'w90pov'
+            ]
+        if '@2.999.999:' in self.spec:
```
Why do these version checks overlap? I think what you want is:
```python
if '@:2' in self.spec:
    ...
else:
    ...
```

adamjstewart commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -26,9 +31,34 @@ class Wannier90(MakefilePackage):
+        return targets
+
+    def url_for_version(self, version):
+        maj_ver = version.up_to(1)
+        if (maj_ver > Version('2')):
```
```suggestion
        if version > Version('2'):
```

adamjstewart commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -76,10 +174,17 @@ def install(self, spec, prefix):
+            if spec.satisfies('@3:'):
+                inst = ['libwannier.a', 'libwannier.so']
+            else:
+                inst = ['libwannier.so']
```
Is a `.so` file also installed on macOS, or is it a `.dylib` file?

downloadico replied:
@adamjstewart: good catch!  I'm guessing you're right on MacOS!
I'm not sure if have a good MacOS box to test on!
"if self.spec.satisfies('platform=darwin'): ..."
Would this work to test if I'm on MacOS?

becker33 replied:
Yes, that's the right test for whether you're on macos.

downloadico committed:
incorporated bbecker's suggestion for making the strings less ugly!

downloadico committed:
incorporated bbecker's suggestion to fix the logic for picking which
"DFT flavor" configure argument.
If the wannier variant is enabled, it passes --with-dft-flavor=wannier90
to configure, otherwise it passes --with-dft-flavor=atompaw+libxc to configure

downloadico committed:
Changed to using plain strings

downloadico committed:
Fixed version tests

downloadico committed:
incorporated @adamjstewart's fix for testing if the major version is > 2

downloadico committed:
incorporated @adamjstewart's fix to check if mpi is enabled and
only set the MPIFC variable if it is.

downloadico committed:
Merge branch 'abinit+wannier90-fix' of https://github.com/downloadico/spack into abinit+wannier90-fix

downloadico committed:
Update var/spack/repos/builtin/packages/wannier90/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

downloadico committed:
Only set MPIFC if '+mpi' is set

downloadico committed:
incorporated fixes from @adamjstewart including:
	- using the string=True argument to filter_file (and removed the unneeded
 	  escapes)
	- changing the url to the github location
	- fixing the version checks
	- building a libwannier.dylib on darwin

downloadico committed:
Merge branch 'abinit+wannier90-fix' of https://github.com/downloadico/spack into abinit+wannier90-fix

Conflicts:
	var/spack/repos/builtin/packages/wannier90/package.py

downloadico committed:
incorporated fixes suggested by @adamjstewart including:
	- using the string=True argument to filter_file and cleaned up the escapes
	- only pass the MPIFC argument to configure when '+mpi' is set
	- chaned the url to the github site for Wannier090
	- fixed the version checks
	- build a 'libwannier.dylib' file when building the shared variant on darwin

becker33 commented on var/spack/repos/builtin/packages/abinit/package.py:
```diff
@@ -95,12 +100,26 @@ def configure_args(self):
+                 .format(spec['wannier90'].prefix.modules))
+            oapp('--with-wannier90-bins={0}'
+                 .format(spec['wannier90'].prefix.bin))
+            if '+mpi' in spec:
+                oapp('MPIFC={0}/mpifc'.format(spec['mpi'].prefix.bin))
```
Can you move this to the same `+mpi` block as the other MPI-specific changes below?

downloadico replied:
probably; I don't think the order of the configure args are important....

becker33 commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -76,10 +176,20 @@ def install(self, spec, prefix):
-        )
+        inst = []
+        if '+shared' in spec:
+            if 'arch=Darwin' in spec:
```
```suggestion
            if spec.satisfies('platform=darwin'):
```

downloadico committed:
Update var/spack/repos/builtin/packages/wannier90/package.py

Co-authored-by: Greg Becker <becker33@llnl.gov>

downloadico committed:
moved a configure argument from it's own '+mpi' check to under the lower one

downloadico commented:
I'm getting strange failures related to, I think python versions (on the test machine that checks PRs)

adamjstewart commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -61,10 +89,82 @@ def edit(self, spec, prefix):
+                              'src/Makefile.2'), string=True)
+
+                filter_file(
+                    ' '.join(['$(AR) $(ARFLAGS)',
+                              '$(LIBRARY) $(OBJS2) $(OBJS)']),
+                    ' '.join(['$(MPIF90) $(FCOPTS) -shared -o '
+                              '$(LIBRARY) $(OBJS2) $(OBJS) $(LIBS)']),
```
I'm pretty sure you can do something like this:
```suggestion
                    '$(AR) $(ARFLAGS) '
                    '$(LIBRARY) $(OBJS2) $(OBJS)',
                    '$(MPIF90) $(FCOPTS) -shared -o '
                    '$(LIBRARY) $(OBJS2) $(OBJS) $(LIBS)',
```
Try that out and see if it parses.

downloadico replied:
It looks like it does!

adamjstewart replied:
Can you do this for the other `' '.join([...])` usages too?

downloadico committed:
Update var/spack/repos/builtin/packages/wannier90/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

downloadico committed:
Cleaned up syntax as suggested by @adamjstewart
It looks *so much better* now!  Thanks!

adamjstewart commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -5,6 +5,7 @@
 
 import inspect
 import os.path
+from llnl.util.filesystem import find
```
You shouldn't need to import this, it should already be present

downloadico replied:
D'oh! :)  I'll fix that soon!

downloadico replied:
Got it!

adamjstewart commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -61,10 +89,81 @@ def edit(self, spec, prefix):
             filter_file(key, value, self.makefile_name)
 
-    def install(self, spec, prefix):
+        if '+shared' in self.spec:
+            if self.spec.satisfies('@:2'):
```
If you want more space, you can either combine these in a single if-statement:
```python
if spec.satisfies('@:2 +shared'):
```
or
```python
if '@:2 +shared' in spec:
```
or you can exit when this isn't needed:
```python
if not spec.satisfies('@:2 +shared'):
    return
```
That way the line length issue isn't as bad.

downloadico replied:
I find the first two to be clearer as to the intention, but I'm a python noob...  Thanks for the suggestions!

adamjstewart commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -61,10 +89,81 @@ def edit(self, spec, prefix):
+        if '+shared' in self.spec:
+            if self.spec.satisfies('@:2'):
+                # this is to build a .shared wannier90 library
+                if self.spec.satisfies('platform=darwin'):
+                    filter_file('LIBRARY = ../../libwannier.a',
+                                'LIBRARY = ../../libwannier.dylib',
+                                join_path(self.stage.source_path,
+                                          'src/Makefile.2'))
+                else:
+                    filter_file('LIBRARY = ../../libwannier.a',
+                                'LIBRARY = ../../libwannier.so',
+                                join_path(self.stage.source_path,
+                                          'src/Makefile.2'))
```
```suggestion
                filter_file('LIBRARY = ../../libwannier.a',
                            'LIBRARY = ../../libwannier.' + dso_suffix,
                            join_path(self.stage.source_path, 'src/Makefile.2'))
```

adamjstewart commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -61,10 +89,81 @@ def edit(self, spec, prefix):
+                                 'geninterp.o']
+                filter_file(''.join(['OBJS_POST  = ', ' '.join(objs_post)]),
+                            ''.join(['OBJS_POST  = ',
+                                     ' '.join(new_objs_post)]),
```
You should be able to remove these joins too

downloadico replied:
Thanks (and thank goodness!)

adamjstewart commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -76,10 +175,20 @@ def install(self, spec, prefix):
-        )
+        inst = []
+        if '+shared' in spec:
+            if spec.satisfies('platform=darwin'):
+                inst.append('libwannier.dylib')
+            else:
+                inst.append('libwannier.so')
```
```suggestion
            inst.append('libwannier.' + dso_suffix)
```

downloadico replied:
Thanks!  I've changed this.

downloadico committed:
removed unneeded import of 'find' from 'llnl.util.filesystem' package
as suggested by @adamjstewart

downloadico committed:
Update var/spack/repos/builtin/packages/wannier90/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

downloadico committed:
incorporated changes from @adamjstewart
changed check to "if '@:2 +shared' in spec:" instead of a nested check of '@:2' and
'+shared'
removed unneeded joins used in filter_file and spliced the list of objs directly into
the filter_file call
used the dso_suffix instead of testing for darwin to determine the name of the
shared library

downloadico committed:
removed whitespace from blank line

becker33 commented:
@downloadico please don't mention people by github handle in commit messages, given the number of PRs in this repo it can become a major source of spam for people like @adamjstewart who review lots of PRs.

downloadico commented:
Thanks!  Will do!

Jim

On Wed, 15 Jul 2020, Greg Becker wrote:

> @downloadico please don't mention people by github handle in commit messages, given the number of PRs in this repo it can become a major source of spam for people like @adamjstewart who review lots of PRs.
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/spack/spack/pull/17417#issuecomment-658963732

adamjstewart commented on var/spack/repos/builtin/packages/wannier90/package.py:
```diff
@@ -61,10 +88,68 @@ def edit(self, spec, prefix):
+                        'kpath.o kslice.o boltzwann.o geninterp.o',
+                        join_path(self.stage.source_path,
+                                  'src/Makefile.2'))
+            filter_file('../../wannier90.x: .*',
```
Is this `.*` actually present in the source code, or do you want to expand to match any string? If it's the latter, you'll need to remove the `string=True` below.

downloadico replied:
Thank you!  You are correct, it should be treated as a regexp!

downloadico replied:
Oops!  I introduced a bug with some missing whitespace!  I will fix it soon and push that fix!  (the build step failed and I'm fairly certain this is the culprit)

downloadico committed:
fixed bug with '../../wannier90.x: .*' not being treated as a regexp.  Thanks Adam!

downloadico committed:
fixed missing whitespace when modifying Makefile.2

adamjstewart commented:
@becker33 can you re-review?
